### ISSUE TITLE
[cryptolib] API cleanup

### ISSUE THREAD

**Issue by jadephilipoom** [lowRISC/opentitan#19549]
### Description

This is a tracking issue for a few changes I'd like to make to the cryptolib API before too much stuff starts relying on it and causing compatibility issues. Most of these changes we can live without, but I'm prioritizing this work because it will be much easier to do it now than later and the cryptolib implementation work is about to finish a major milestone, so it would be a good time to fix some tech debt.

- [x] Change ECDSA/RSA signature operations to take message digests instead of raw messages as input (it's possible to build the raw-message interface out of this, but not possible to do the reverse, so this gives the caller more flexibility; note that this is not applicable to Ed25519 since it has a fixed hash function that must be used in a specific way internally)
- [x] Update the AES IV on return (since we only have one-shot AES, this makes it easier for callers to build their own streaming version)
- [x] Make the usage of pointers/const more uniform across API functions
- [x] Introduce word buffers and use them for signatures, digests, MACs, ciphertexts, and DRBG input/output (basically everything except message inputs for hash functions/signatures)
- [x] Separate SHAKE/cSHAKE functions (this was already done in the API doc but hasn't been reflected in the code yet)
- [x] Change the naming of aes_kwp functions to "wrap/unwrap" instead of "encrypt/decrypt" (to match the spec and also avoid someone trying to use it to encrypt normal data)
- [ ] Review FIPS requirements for avoiding using keys across different modes and ensure we have enough key types to prevent cross-use (sometimes these are really specific, e.g. you can't use the same key with different padding schemes)
- [x] Change the representation of hardware-backed keys to store diversification data in the keyblob to save some storage space and representation complexity
- [ ] Adjust async interface if needed to use OTBN's checksum register (see https://github.com/lowRISC/opentitan/issues/19173)

I've also been considering the following changes, but haven't discussed them much yet, so opinions very much welcome:

- [x] Change key transport interface? (replace `blinded_to_unblinded`/`unblinded_to_blinded`/`build_unblinded`/`build_blinded` with `import_blinded`/`import_unblinded`/`export_blinded`/`export_unblinded`, so we don't have unblinded keys that are secretly blinded keys and the usage is a little clearer)
- [x]  Remove specialized RSA/ECC key/signature structs and replace them with simply blinded or unblinded keys that then get unpacked
- [x] Make it possible to have a `NULL` for custom Weierstrauss parameters, or define constants for named curves, so we don't have to fully populate a domain parameters struct with empty buffers every time we want to use a named curve
- [x] ~Remove synchronous functions for asynchronous operations (these seem a little redundant to me; they just call the async versions internally, and removing them would save code size and API surface area)~ (edit: removed due to discussion below)
- [x] ~Remove one-shot functions for operations that have an init/update/final interface (same rationale as above for async)~ (edit: removed due to discussion below)
- [x] Change AES-GCM interface to have a more typical streaming mode instead of exposing GHASH/GCTR (I advocated for the current interface previously because streaming GCM can encourage decrypting without checking the tag, but there is enough to GCM besides GHASH/GCTR that I'm now concerned about integration and hardening complexity with the current version)
- [x] Either separate HMAC/KMAC keygen or unify AES/HMAC/KMAC keygen into one "generate symmetric key" function (currently we have AES keygen and a unified HMAC/KMAC keygen, which seems a little inconsistent to me)
- [ ] Change `len` terminology to be clearer about word vs. byte length.

CC @moidx @y-srini

**Comment by lschuermann** [comment]
> Introduce word buffers and use them for signatures, digests, MACs, ciphertexts, and DRBG input/output (basically everything except message inputs for hash functions/signatures)

Just to be sure: these word-buffers will continue to expect their respective underlying data to be provided in big-endian byte ordering, right? E.g., looking at the HMAC functest, a key of `0x00010203...` would be encoded into a word-buffer as `[0x03020100, ...]`, assuming a 32-bit little-endian target architecture.

This seems like an important property to keep and codify (if it isn't already). For instance, Tock's in-kernel cryptography APIs will use byte-slices to pass around data and key material; we would want to avoid any spurious allocations and expensive operations for byte-order conversions.

**Comment by lschuermann** [comment]
> Remove synchronous functions for asynchronous operations (these seem a little redundant to me; they just call the async versions internally, and removing them would save code size and API surface area)

These blocking helpers do make it easier to invoke cryptographic operations on fundamentally synchronous code paths (for instance by avoiding manual allocation of a context struct, etc.). Also, if CryptoLib were to be executed in an isolated context (such as in user-mode), they may reduce the number of context switches required for a given operation.

On the other hand, at least in the context of Tock, most internal cryptography-APIs are fundamentally asynchronous and would primarily make use of the asynchronous functions. Would it be possible to optionally provided these methods (e.g., controlled through a preprocessor macro)?

**Comment by jadephilipoom** [comment]
> Just to be sure: these word-buffers will continue to expect their respective underlying data to be provided in big-endian byte ordering, right? 

Right, all of these word buffers will maintain the property that, read byte-by-byte, they produce the expected byte stream for the algorithm even if that byte stream is big-endian under the hood.

The motivation for introducing the word buffers is that passing bytes around (including over the bus) is risky from an SCA perspective, since there are only 256 possibilities for the value and attackers can analyze it easily. Of course, it's also a bit faster to use word-aligned writes, but the SCA concern is my primary motivation.

**Comment by jadephilipoom** [comment]
> On the other hand, at least in the context of Tock, most internal cryptography-APIs are fundamentally asynchronous and would primarily make use of the asynchronous / split functions. Would it be possible to optionally provided these methods (e.g., controlled through a preprocessor macro)?

The asynchronous operations are only provided for things that are likely to take longer than 5-10ms to complete, e.g. ECC and RSA; quick operations will have the synchronous interface only. My thinking here is that at the point when you're running one of these long operations, the context switch is pretty cheap in comparison. It would also potentially be possible to make them optional, or to add them later on if there's a good justification (easier to remove them now and add them if we need them than to keep them and remove them later when other things might be calling them).

**Comment by y-srini** [comment]
> Change ECDSA/RSA signature operations to take message digests instead of raw messages as input 
Agree ; (We discussed it before, just adding the notes from the discussion)
- The API would have an identifier for the hash function for digest
- Either an identifier for raw message/digest in the API; or two separate APIs for raw message & digest - What are your thoughts?
While we are at it, would be interesting to revisit the supported the hash modes.

**Comment by y-srini** [comment]
> Change key transport interface? (replace blinded_to_unblinded/unblinded_to_blinded/build_unblinded/build_blinded with import_blinded/import_unblinded/export_blinded/export_unblinded, so we don't have unblinded keys that are secretly blinded keys and the usage is a little clearer)

Agree. When this change is made, a reminder to - add a note on the non-exportable keys (from crypto_key_config_t) for the export blinded keys.

**Comment by y-srini** [comment]
> Remove synchronous functions for asynchronous operations
> Remove one-shot functions for operations that have an init/update/final interface

Agree on the redundancy part. On the other hand for the crypto lib user, might be convenient/easier to use these one-shot APIs. I would lean on keeping it followed by controlling it through a preprocessor macro as suggested.

**Comment by y-srini** [comment]
> Either separate HMAC/KMAC keygen or unify AES/HMAC/KMAC keygen into one "generate symmetric key" function
I guess, unifying them sounds good, with an identifier for key-mode.

**Comment by jadephilipoom** [comment]
> * Either an identifier for raw message/digest in the API; or two separate APIs for raw message & digest - What are your thoughts?
>   While we are at it, would be interesting to revisit the supported the hash modes.

I think it would help reduce code size and API surface area to just have one single interface that accepts a digest. If we do need the other version later, it'll be easier to add it than to remove it. And I think in this case we could take `hash_mode_t` as the digest parameter and support any hash function that is an acceptable length and compatible with the signature scheme.

>> Remove synchronous functions for asynchronous operations
>> Remove one-shot functions for operations that have an init/update/final interface

> These blocking helpers do make it easier to invoke cryptographic operations on fundamentally synchronous code paths (for instance by avoiding manual allocation of a context struct, etc.). Also, if CryptoLib were to be executed in an isolated context (such as in user-mode), they may reduce the number of context switches required for a given operation.

> For the crypto lib user, might be convenient/easier to use these one-shot APIs. I would lean on keeping them ; otherwise to control it through a pre-processor macro, as @lschuermann suggested.

Since both of you agree that these synchronous or one-shot wrappers are useful, let's leave them in for now. It might make sense at some point to offer different compile-time packaging for cryptolib, e.g. "lite" if you want to save on code size at the expense of performance/API convenience. In that case we could filter out these functions for those targets which prioritize code size.

**Comment by johannheyszl** [comment]
IIUC, the tasks have been completed
